File: src/api/wg/wg.py
import os
import tempfile
import uuid
from typing import List

from .executor import execute
from .models import PeerUsage, KeyPair


def __parse_val(value: str, fn: type = None):
    if value == '(none)':
        return None
    if fn:
        return fn(value)
    return value.strip()


def create_keypair() -> KeyPair:
    (private_fd, private_key_path) = tempfile.mkstemp(prefix='wg', suffix=f'{uuid.uuid4().hex}')
    (public_fd, public_key_path) = tempfile.mkstemp(prefix='wg', suffix=f'{uuid.uuid4().hex}')
    try:
        execute(f'wg genkey > {private_key_path} && '
                f'wg pubkey < {private_key_path} > {public_key_path}', shell=True)
        with open(private_fd, 'r') as rd:
            private_key = rd.read()
        with open(public_fd, 'r') as rd:
            public_key = rd.read()
    finally:
        os.remove(private_key_path)
        os.remove(public_key_path)
    return KeyPair(public_key, private_key)


def get_peer_ips(interface: str) -> List[str]:
    output, _ = execute(f'wg show {interface} allowed-ips')
    data = []
    if output and len(output) > 0:
        for line in output.split('\n'):
            l = line.split('\t')
            if len(l) != 2:
                continue
            peer, ip = l[0], l[1]
            data.append(ip[:ip.index('/')])
    return data


def add_conf(interface: str, config: str):
    (fd, config_path) = tempfile.mkstemp(prefix='wg.config', suffix=f'{uuid.uuid4().hex}')
    try:
        with open(fd, 'w') as wd:
            wd.write(config)
        execute(f'wg addconf {interface} {config_path}')
    finally:
        os.remove(config_path)


def save_conf(interface):
    execute(f'wg-quick save {interface}')


def remove_conf(interface, key):
    execute(f'wg set {interface} peer {key} remove')


def dumps(interface) -> List[PeerUsage]:
    output, _ = execute(f'wg show {interface} dump')
    peers: List[PeerUsage] = []
    if not output or len(output) == 0:
        return peers
    lines = output.split('\n')[1:]
    for line in lines:
        l = line.split('\t')
        if len(l) != 8:
            continue
        peer = PeerUsage(
            public_key=__parse_val(l[0]),
            endpoint=__parse_val(l[2]),
            allowed_ips=__parse_val(l[3]),
            latest_handshake=__parse_val(l[4], int),
            rx=__parse_val(l[5], int),
            tx=__parse_val(l[6], int),
            persistent_keepalive=__parse_val(l[7]))
        peers.append(peer)
    return peers

File: api/wg.go
package api

import (
	"fmt"
	"io/ioutil"
	"net"
	"os"
	"strings"
	"sync"
)

const tmpDir = "/tmp"

type wireguard struct {
	sync.Mutex // sync all wireguard-tools call
	api        *Api
}

func (api *Api) initWgApi() {
	wg := new(wireguard)
	wg.api = api
	api.wg = wg
}

func (wg *wireguard) saveConf(itf string) error {
	wg.Lock()
	defer wg.Unlock()
	if err := execCmd("wg-quick save %s", itf); err != nil {
		return fmt.Errorf("save conf exec %v", err)
	}
	wg.api.log.Printf("save conf success")
	return nil
}

func (wg *wireguard) addConf(itf string, conf string) error {
	wg.Lock()
	defer wg.Unlock()
	api := wg.api
	var err error
	file, err := ioutil.TempFile(tmpDir, "wgc.*.conf")
	if err != nil {
		return fmt.Errorf("add conf tempfile %v", err)
	}
	defer os.Remove(file.Name())
	if _, err = file.Write([]byte(conf)); err != nil {
		return fmt.Errorf("add conf write tempfile %v", err)
	}
	if err = execCmd("wg addconf %s %s", itf, file.Name()); err != nil {
		return fmt.Errorf("add conf (%s) exec error %v", conf, err)
	}
	api.log.Printf("add conf (%s) success", conf)
	return nil
}

func (wg *wireguard) removeConf(itf string, publicKey string) {
	wg.Lock()
	defer wg.Unlock()
	if err := execCmd("wg set %s peer %s remove", itf, publicKey); err != nil {
		wg.api.log.Errorf("remove conf (%s, %s) exec %v", itf, publicKey, err)
	}
	wg.api.log.Printf("remove conf success")
}

func (wg *wireguard) getPublicKey(itf string) (string, error) {
	wg.Lock()
	defer wg.Unlock()
	output, err := execCmdGetOutput("wg show %s public-key", itf)
	if err != nil {
		return "", fmt.Errorf("wg show %s public-key - exec %v", itf, err)
	}
	for _, s := range strings.Split(output, "\n") {
		if strings.TrimSpace(s) != "" {
			return s, nil
		}
	}
	return "", fmt.Errorf("wg show %s public-key - empty output", itf)
}

func (wg *wireguard) getPeerIPs(itf string) ([]net.IP, error) {
	wg.Lock()
	defer wg.Unlock()
	output, err := execCmdGetOutput("wg show %s allowed-ips", itf)
	if err != nil {
		return nil, fmt.Errorf("wg.getPeerIPs(%s) - exec %v", itf, err)
	}
	var ips = make([]net.IP, 0)
	lines := strings.Split(output, "\n")
	for _, line := range lines {
		parts := strings.Split(line, "\t")
		if len(parts) != 2 {
			continue
		}
		_, ip, err := net.ParseCIDR(parts[1])
		if err != nil {
			wg.api.log.Errorf("wg.getPeerIPs(%s) - line (%s) - ParseCIDR %v", itf, line, err)
			continue
		}
		ips = append(ips, ip.IP)
	}
	return ips, nil
}
func (wg *wireguard) dumps(itf string) ([]peerUsage, error) {
	wg.Lock()
	defer wg.Unlock()
	output, err := execCmdGetOutput("wg show %s dump", itf)
	if err != nil {
		return nil, fmt.Errorf("wg.dumps(%s) - exec %v", itf, err)
	}
	var peers = make([]peerUsage, 0)
	for _, line := range strings.Split(output, "\n") {
		parts := strings.Split(line, "\t")
		if len(parts) != 8 {
			continue
		}
		peer := peerUsage{
			parseStr(parts[0]),
			parseStr(parts[2]),
			parseStr(parts[3]),
			parseInt64(parts[4]),
			parseInt64(parts[5]),
			parseInt64(parts[6]),
			parseStr(parts[7]),
		}
		peers = append(peers, peer)
	}
	return peers, nil
}

func (wg *wireguard) isLinkUp(itf string) bool {
	f, err := os.Open(fmt.Sprintf("/sys/class/net/%s/carrier", itf))
	if err != nil {
		return false
	}
	defer f.Close()
	return true
}
